README.md
# spora: Streamlined Phylogenomic Outbreak Report Analysis

![example workflow](https://github.com/matt-sd-watson/spora/actions/workflows/main.yml/badge.svg)

snakemake and Python integrated workflow for intermediate file generation for COVID outbreak analysis

## Installation

```
git clone https://github.com/matt-sd-watson/spora.git
conda env create -f ncov_spora/environments/environment.yml
conda activate ncov_spora
cd spora
pip install . 
```

## Updating

```
conda activate ncov_spora
cd ~/spora
git checkout main
git pull
pip install . 
```

## Usage
```
usage: 
    	spora -c <config.yaml> 
    	OR
    	spora --focal_list ...<input args>

spora: Streamlined Phylogenomic Outbreak Report Analysis

optional arguments:
  -h, --help            Show the help output and exit.
  -c CONFIG, --config CONFIG
                        Input config file in yaml format, all command line arguments can be passed via the config file.
  -f FOCAL_SEQS, --focal-sequences FOCAL_SEQS
                        Input .txt list or multi-FASTA focal samples for outbreak. Required
  -b BACKGROUND_SEQS, --background-sequences BACKGROUND_SEQS
                        Optional input .txt list or multi-FASTA background samples to add to analysis
  -m MASTER_FASTA, --master-fasta MASTER_FASTA
                        Master FASTA of genomic sequences to select from. Required if either --focal-sequences or --background-sequences are not supplied in FASTA format
  -o OUTDIR, --output-directory OUTDIR
                        Path to the desired output directory. If none is provided, a new folder named spora will be created in the current directory
  -r REFERENCE, --reference REFERENCE
                        .gb file containing the desired COVID-19 reference sequence. Required
  -p PREFIX, --prefix PREFIX
                        Prefix string to label all output files. Default: outbreak
  -t NTHREADS, --nthreads NTHREADS
                        Number of threads to use for processing. Default: 2
  -s, --snps-only       Generate a snps-only FASTA from the input FASTA. Default: False
  -rn, --rename         Rename the FASTA headers to be compatible with NML standards. Default: False
  -nc NAMES_CSV, --names-csv NAMES_CSV
                        Use the contents of a CSV to rename the input FASTA. Requires the following column headers: original_name, new_name
  -ncs, --no-constant-sites
                        Do not enable constant sites to be used for SNPs only tree generation. Default: Enabled
  -fi, --filter         Filter both the focal and background sequences based on genome completeness and length. Default: Not enabled
  -gc GENOME_COMPLETENESS, --genome-completeness GENOME_COMPLETENESS
                        Integer for the minimum genome completeness percentage for filtering. Default: 90
  -gl GENOME_LENGTH, --genome-length GENOME_LENGTH
                        Integer for the minimum genome length for filtering. Default: 29500
  -rp, --report         Generate a summary output report for the spora run. Default: Not enabled
  -v, --version         Show the current spora version then exit.
```

## Documentation

More detailed documentation for spora usage and functionality can be found [here](docs/0-OVERVIEW.md)

## Acknowledgments

Inspiration for code structure and design for spora was inspired by [pangolin](https://github.com/cov-lineages/pangolin) and [civet](https://github.com/artic-network/civet), and minor code blocks were adopted from these software.

The **Background** section in the documentation describing outbreak definitions was written by Mark Horsman. 


CHANGELOG.md
# spora Changelog

## Version 0.1.0, 06-10-21
- Initial workflow
- Required inputs: focal sequence list (.txt), COVID-19 reference (.gb),
 master fasta sequences from which to subset (.fa, .fasta)
 - Optional features include output directory, prefix, SNPS only analysis,
 ability to rename headers to custom or match NML specifications,
 and toggling constant sites for SNPs only analysis
 
 
## Version 0.2.0, 02-11-21
 - focal and background filtering step can now be used with --filter,
 where the sequences can be filtered based on genome completeness and
 genome length. The defaults are set to 90% genome completeness and
 a genome length of 29500 for ncov
 
 
## Version 0.3.0, 15-11-21
 - focal and background sequences can now be supplied as either .txt list files 
 of sample names, or as FASTA files containing the actual sequences. outbreaker 
 will assess the file ending and evaluate whether to create subsets based on the 
 names (if .txt is passed) or use the FASTA given
 
 
## Version 0.4.0, 19-11-21
 - if a previous run of iqtree2 for the SNPs only tree is detected, the console will
 allow the user to select how to proceed. Currently the only possible selection is to confirm
 overwriting the existing SNPs only tree analysis, or outbreaker will exit. Requires a new environment
 installation as the logic uses click through pip. 
 
 
## Version 0.5.0, 29-11-21
 - outbreaker now creates an output summary report summarizing the number of input sequences, SNP distance patterns relative to the focal sequences inputs, as well as basic rendering of the phylogenetic trees. The report is flexible: it will modify the outputs accordingly if either background sequences are not supplied or SNPs only analysis is not conducted. requires a majoy environment upgrade, specifically with regards to R and Bioconductor dependencies through conda
 
  
## Version 0.6.0, 01-12-21
 - outbreaker now uses snipit to create a graphic of SNP diversity relative to the reference, and includes it in the summary report. This requires an environment upgrade to add snipit through pip
 - The summary report can now handle the recognition of focal and background sequences when samples are renamed through ```--rename```.
 - Still requires teting to verify that ```--rename``` with ```--names_csv``` works in the same way as above


## Minor Version 0.6.1, 02-12-21
 - The output summary report is now an optional output to avoid rendering a large report by default with a
   very large dataset. This can be enabled with ```--report```. By default the report is not generated.
  
   
## Minor Version 0.6.2, 07-12-21
 - ```outbreaker -v``` or ```outbreaker --version``` will now show the current version, then exit. 
 - In the summary report, the trees are automatically scaled with xlim to ensure that the tree labels are always visible, no matter the length of the branch
 - In the summary report, a treescale was added for the trees that renders in either the bottom right or top right, depending on where the longest branches are. This was done to avoid having the treescale overlapping a tip lab. 
  
  
## Minor Version 0.6.3, 10-12-21
 - ```--names-csv``` is now compatible with the summary report 
 - ```--filter``` and ```--names-csv``` are now compatible. Previously,, if any sequences were filtered and a names CSV was supplied, there would be an error.
 - The reference ncov sequence that is compatible with MAFFT (augur align) is now included under ```/data/reference/``` as ```ncov_reference.gb``` 
 
## Minor Version 0.6.4, 17-02-22

 - outbreaker now retains all sequences if ```--names-csv``` is used for renaming and not all sequences are contained in the CSV
 - updates to the renaming behavior to be compatible with fastafurious v1.2.0 (additional warning messages)
 
## Minor Version 0.6.5, 28-04-22 (Patch)
 - Change behaviour of renaming when no CSV is supplied. Will now use the prefix for the run to generate new names with alphanumerical sequential order (i.e. prefix_1, prefix_2) and will output the name matches as a CSV file
 - Above fix changes fixes the error in the SNP distance plot in the HTML report when rename is used but no names CSV is supplied
 - Addition of pytests in the CI/CD workflow
 
 ## Version 0.7.0, 15-06-22
 
 - outbreaker is renamed to spora to be compatible with PyPi
 - version change to dplyr dependency for conda env
 - Addition of CD GA workflow to pudblish to PyPi on release
 


spora/main.py
from spora import __version__
import spora.init_defaults as defaults

import os
import sys
import argparse
import snakemake
import yaml

cwd = os.getcwd()
thisdir = os.path.abspath(os.path.dirname(__file__))


def isFasta(input):
    return input.endswith(('.fa', '.fasta', '.FA', '.FASTA'))

# set mandatory input args from either config or CLI
# mandatory = set(["focal_list", "reference", "master_fasta"])

def get_primary_snakefile(thisdir):
    snakefile = os.path.join(thisdir, 'workflows', 'spora.smk')
    if not os.path.exists(snakefile):
        print(f'Error: cannot find Snakefile at {snakefile}\n Check installation\n')
        sys.exit(-1)
    return snakefile


def main(sysargs = sys.argv[1:]):
    parser = argparse.ArgumentParser(add_help=False,
                                    description="spora: Streamlined Phylogenomic Outbreak Report Analysis",
                                    usage='''
    \tspora -c <config.yaml> 
    \tOR
    \tspora --focal_list ...<input args>''')

    parser.add_argument('-h', "--help", action="help",
                        help="Show the help output and exit.",
                        dest="help")

    parser.add_argument('-c', "--config", action="store",
                         help="Input config file in yaml format, all command line arguments can be passed via the config file.",
                         dest="config")

    parser.add_argument('-f', "--focal-sequences", action="store",
                        help="Input .txt list or multi-FASTA focal samples for outbreak. Required",
                        dest="focal_seqs", default="")

    parser.add_argument('-b', "--background-sequences", action="store",
                        help="Optional input .txt list or multi-FASTA background samples to add to analysis",
                        dest="background_seqs", default="")

    parser.add_argument('-m', "--master-fasta", action="store",
                        help="Master FASTA of genomic sequences to select from. Required if either --focal-sequences "
                             "or --background-sequences are not supplied in FASTA format",
                        dest="master_fasta", default="")

    parser.add_argument('-o', "--output-directory", action="store",
                        help="Path to the desired output directory. If none is provided, "
                             "a new folder named spora will be created in the current directory",
                        dest="outdir", default="")

    parser.add_argument('-r', "--reference", action="store",
                        help=".gb file containing the desired COVID-19 reference sequence. Required",
                        dest="reference", default="")

    parser.add_argument('-p', "--prefix", action="store",
                        help="Prefix string to label all output files. Default: outbreak",
                        dest="prefix", default="outbreak")

    parser.add_argument('-t', "--nthreads", action="store",
                        help="Number of threads to use for processing. Default: 2",
                        dest="nthreads", default=2, type=int)

    parser.add_argument('-s', "--snps-only", action="store_true",
                        help="Generate a snps-only FASTA from the input FASTA. Default: False",
                        dest="snps_only")

    parser.add_argument('-rn', "--rename", action="store_true",
                        help="Rename the FASTA headers to be compatible with NML standards. Default: False",
                        dest="rename")

    parser.add_argument('-nc', "--names-csv", action="store",
                        help="Use the contents of a CSV to rename the input FASTA. Requires the following "
                             "column headers: original_name, new_name",
                        dest="names_csv", default="")

    parser.add_argument('-ncs', "--no-constant-sites", action="store_false",
                        help="Do not enable constant sites to be used for SNPs only tree generation. Default: Enabled",
                        dest="const_sites")

    parser.add_argument('-fi', "--filter", action="store_true",
                        help="Filter both the focal and background sequences based on genome "
                             "completeness and length. Default: Not enabled",
                        dest="filter")

    parser.add_argument('-gc', "--genome-completeness", action="store",
                        help="Integer for the minimum genome completeness percentage for filtering. Default: 90",
                        dest="genome_completeness", default=90, type=int)
    parser.add_argument('-gl', "--genome-length", action="store",
                        help="Integer for the minimum genome length for filtering. Default: 29500",
                        dest="genome_length", default=29500, type=int)

    parser.add_argument('-rp', "--report", action="store_true",
                        help="Generate a summary output report for the spora run. Default: Not enabled",
                        dest="report")

    parser.add_argument('-v', "--version", action="version",
                        help="Show the current spora version then exit.",
                        version=f"This is spora: v{__version__}")



    if len(sysargs) < 1:
        parser.print_help()
        sys.exit(0)
    else:
        args = parser.parse_args(sysargs)

    if args.config:
        # if args are passed in a config file, they overwrite any existing args or others passed by CLI

        all_valid_inputs = True
        with open(args.config, "r") as f:
            args_to_dict = defaults.load_yaml(f)

        config = defaults.setup_config_dict(cwd, args.config)

        if isFasta(str(config["focal_seqs"])):
            # if both focal and background are fasta, no need for master file
            if config["background_seqs"] == '' or isFasta(str(config["background_seqs"])):
                mandatory = set(["focal_seqs", "reference"])
            # if focal is fasta but background is not, keep master file
            elif config["background_seqs"] != '' or not isFasta(str(config["background_seqs"])):
                mandatory = set(["focal_seqs", "reference", "master_fasta"])
        else:
            mandatory = set(["focal_seqs", "reference", "master_fasta"])

        current_keys = set([key for key in args_to_dict])
        if len(current_keys.intersection(mandatory)) != len(mandatory):
            sys.stderr.write(f'One of more required parameters were not passed in the config:\n{mandatory}\n')
            sys.exit(-1)

        all_valid_inputs = True
        for key in args_to_dict:
            if args_to_dict[key] == "" or args_to_dict[key] is None and key in mandatory:
                all_valid_inputs = False
                sys.stderr.write(f'The following config parameter was passed as empty. It is required: {key}\n')

        if not all_valid_inputs:
            sys.stderr.write(f'ERROR: Please review the config inputs\n')
            sys.exit(-1)

    else:
        # if the args are not passed in a config file, create a config dictionary for snakemake
        config = vars(args)

        if isFasta(str(config["focal_seqs"])):
            # if both focal and background are fasta, no need for master file
            if config["background_seqs"] == '' or isFasta(str(config["background_seqs"])):
                mandatory = set(["focal_seqs", "reference"])
            # if focal is fasta but background is not, keep master file
            elif config["background_seqs"] != '' or not isFasta(str(config["background_seqs"])):
                mandatory = set(["focal_seqs", "reference", "master_fasta"])
        else:
            mandatory = set(["focal_seqs", "reference", "master_fasta"])

        all_valid_inputs = True
        for key in config.keys():
            # If the mandatory key is not passed on the CLI, flag and exit
            if config[key] == '' and key in mandatory:
                all_valid_inputs = False
                sys.stderr.write(f'The following CLI parameter was not passed. It is required: {key}\n')

        if not all_valid_inputs:
            sys.stderr.write(f'ERROR: Please review the command line inputs\n')
            sys.exit(-1)

    snake_file = get_primary_snakefile(thisdir)

    status = snakemake.snakemake(snake_file, printshellcmds=True, forceall=True, force_incomplete=True,
                                 config=config, cores=args.nthreads, lock=False, quiet = True
                                 )

    if status:  # translate "success" into shell exit code of 0
        sys.stderr.write(f"\nspora completed successfully.\n")
        return 0

    sys.stderr.write(f"\nspora did not complete successfully.\n")
    return 1

if __name__ == '__main__':
    main()
